### PR TITLE
feat(reports): Add custom JFR event for report rules

### PR DESCRIPTION
**Pull request by jan-law**
Related #461 
Depends on https://github.com/cryostatio/cryostat-core/pull/118
Depends on https://github.com/cryostatio/cryostat-core/pull/121

Adds the custom JFR event from `InterruptibleReportGenerator` to the Cryostat template

### CODE REVIEW

**Review by andrewazores — COMMENTED** [review]
Did you want to update the `SubprocessReportGenerator` to use the `Interruptible` implementation in this PR too?

**Comment by jan-law** [comment]
Sure, I can update it. I didn't include the change because I thought the `SubprocessReportGenerator` was supposed to stay as is, but now I realize it's probably a good idea to use the interruptible implementation.

**Comment by andrewazores** [comment]
The changes to the implementation are pretty minimal so I think it's safe. If we update it then we can get profiling samples for subprocess generation vs sidecar generation and compare performance between them as directly as possible and see if there are any unexpected discrepancies, so I think that's worth it on its own already.

**Comment by andrewazores** [comment]
The subprocess being forked out does also allow the main Cryostat process to forcibly kill the subprocess if it times out, so the interrupt-ibility isn't really gaining us anything there, though.

**Comment by andrewazores** [comment]
I haven't had any luck getting the event to appear in recordings where I'm using subprocess generation and recording Cryostat itself using its own `cryostat.jfc` template. Not sure if I have just overlooked something or am building something wrong - have you got it to work? I'm using the `jfr` CLI command to examine the JFR files produced by the `cryostat.jfc` recording, so ex. `jfr print foo.jfr | grep RuleEval`

**Comment by jan-law** [comment]
Here's my recording file from yesterday. `jfr print /home/jalaw/Downloads/_deployments_quarkus-run-jar_testing_20220124T173358Z.jfr | grep ReportRuleEval` should return results:
file:///home/jalaw/Downloads/_deployments_quarkus-run-jar_testing_20220124T173358Z.jfr

I'm running into an error when running `./mvnw package` today, which means I'm unable to generate reports right now. Have you seen this error before?
```
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[ERROR] TestEngine with ID 'junit-jupiter' failed to discover tests
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.393 s
[INFO] Finished at: 2022-01-25T15:09:15-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M5:test (default-test) on project cryostat-reports: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/jalaw/cryostat-reports/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] There was an error in the forked process
[ERROR] TestEngine with ID 'junit-jupiter' failed to discover tests
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
[ERROR] TestEngine with ID 'junit-jupiter' failed to discover tests
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:733)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:305)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:265)
```

**Comment by andrewazores** [comment]
Haven't seen that one before and can't reproduce it either, on main or on your PR branch. That's pretty strange looking.

**Comment by jan-law** [comment]
I realized I forgot to push the update -core version change to `pom.xml`.

Update: `./mvnw package` is working for me now after waiting a few minutes. I'm still getting a `Type Error: failed to fetch report` on the frontend though. I'll try rebuilding and see what happens.

**Comment by jan-law** [comment]
> Judging by the filename, that was produced by doing a recording on a sidecar, correct? 

Yes, I started the recording on the sidecar container on http port 10001

> What event template did you use?

I used the modified Cryostat template from this branch

**Comment by andrewazores** [comment]
Okay, I tried doing the same by downloading the Cryostat template as a TARGET template from the Cryostat instance, and then re-uploading it so I could have a CUSTOM template copy that can be applied to other targets. Then, applying that version of the template to start a recording against the sidecar worked and I was able to collect the new event and see them with `jfr print`.

No events when I use subprocess generation, however.

**Comment by andrewazores** [comment]
Ah. D'oh. What a brainfart. Of course we don't get the events captured in the recording when we are using subprocess generation - because the event is not emitted and recorded in the Cryostat process, but in the fork subprocess, so it's a separate JVM which we are not connecting to and not running a recording in. I don't think we have a feasible technical solution to that at this point in time, really, and it isn't worth putting time into on its own.

Maybe the next best thing we can do is add another evaluation event emission that captures the entire report generation time frame from the POV of the process being recorded, rather than the individual rules being evaluated. So for the -reports sidecar, this would be nearly equivalent to just timing the whole API request, but without tracking any of the network latency on either end of the request. For the subprocess generation this event would be recorded by the main Cryostat process, not the subprocess, and to make it as close to an apples-to-apples comparison to the sidecar generation, the event would be started right before the subprocess is forked (after the JFR file has been created on-disk, if needed, for the subprocess to read), and the event ended/committed as soon as the generation completes and flow returns back to the main Cryostat process - so either the subprocess has timed out and the main process has decided to forcibly kill it, or the subprocess completed normally and the main process has stopped waiting on it, but not yet proceeded to read the generated report file and send it out over the network.

Does that make sense? I can provide more specific method calls where I mean the events should be injected if that wasn't well explained.

**Comment by jan-law** [comment]
> the event is not emitted and recorded in the Cryostat process, but in the fork subprocess

Ah, thanks for figuring that out. I was unsuccessfully trying to figure out where the event data goes after the subprocess completes.

> for the -reports sidecar, this would be nearly equivalent to just timing the whole API request

Checking my understanding here, it would look something like this?
```
RemoteReportGenerator {
  get request parameters
  // start event here
  POST generate report
  parse response
  // end event here
  return response
}
```

And for the subprocess report generator, did you mean to record the report generation inside the api handlers, while excluding the time it takes for the `TargetReportGetHandler` to fetch the recording from the JVM, like this?
```
ReportGetHandler {
...
  handleWithValidJwt() {
  ...
  //start event here
  Path report =
                      reportService
                              .get(recordingName)
                              .get(generationTimeoutSeconds, TimeUnit.SECONDS);
  // end event here
  ...
  }
}
```
```
TargetReportGetHandler {
...
  handleWithValidJwt() {
  ...
 reportService
      .get(getConnectionDescriptorFromJwt(ctx, jwt), recordingName)
// create another variable here to split up the .get() chain
//  start event here
      .get(reportGenerationTimeoutSeconds, TimeUnit.SECONDS));
  // end event here
  ...
  }
}
```

**Comment by andrewazores** [comment]
For RemoteReportGenerator we have two options. One is to record the event on Cryostat's side:

```
API request from client to Cryostat
Cryostat pulls JFR from target if recording is active
Cryostat POSTs JFR to remote report generator
-> start event here, immediately after the POST has completed
Report sidecar processes request
Report sidecar writes response to wire
Cryostat starts to receive POST response on wire
-> end event here
Cryostat finishes receiving POST response on wire
Cryostat takes POST response from sidecar and writes that back to the client to complete the original API request
```

This might be a little tricky since you have to know about the request/response being in progress to time the event start/stop properly. I imagine this does mean the event has to be implemented in the reports API handlers. The other option is to just record it on the -reports sidecar side:

```
POST request received and handled by Quarkus
Application code starts running - ReportResource handler called
-> start event here
ReportResource handler calls InterruptibleReportGenerator to do the actual work
Report generation completed or timed out
-> end event here
Send generated report or error status back as POST response
```

This second way is clean and easy to implement, but it does mean that in order to profile the performance then the recording has to actually be applied to the reports sidecar(s), and not to the main Cryostat process.

For subprocess generation, we can only really do it from the main Cryostat process, so it looks a lot like the first way laid out for the RemoteReportGenerator:

```
API request from client to Cryostat
Cryostat pulls JFR from target if recording is active
Cryostat writes JFR to disk if it was an active recording, or uses the archived recording already on disk
-> start event here
Cryostat forks the subprocess report generator, providing it with a local path to the JFR file
Subprocess report generator runs and writes its result to another path, then exits
Cryostat notices that the subprocess has exited, or times out waiting for this to happen
-> end event here
Cryostat uses the path where the subprocess wrote its report to and sends the contents as the reports API response
```

I imagine this is done in SubprocessReportGenerator, basically in the `exec` method.

**Comment by jan-law** [comment]
> This second way is clean and easy to implement, but it does mean that in order to profile the performance then the recording has to actually be applied to the reports sidecar(s), and not to the main Cryostat process.

It seems like applying the recording to the report sidecar would give a better timing comparison. I'm wondering what would happen if you had multiple sidecar replicas though. If I started a recording on all of my replicas, then decided to generate a report, how would I know which replica recording contained the report generation events without manually checking all the recordings? Although I'm not sure what kind of use cases would exist here, I suppose this question becomes relevant if someone needed many replicas at once.

**Comment by andrewazores** [comment]
You wouldn't know which specific replica processed any particular report request, but I think the interesting thing will be to gather aggregate data about report generation performance, so it doesn't matter too much that we are lacking traceability on the request. I think the utility of this rule evaluation event is mostly going to be to help users fine-tune their resource requirements/allocations (and for us to set reasonable defaults that work for most users), so aggregate performance data is what really counts. We have some tuning knobs to turn (CPU and memory allocation) and some input parameters (size of recording in bytes, "complexity" of recording in how many events are enabled), and we want to see how the response time function varies with those inputs and parameters.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "java.configuration.updateBuildConfiguration": "automatic"
+}
```
Comment (andrewazores) [inline .vscode/settings.json:3]: This looks like an editor-specific file - I don't think we want to check it in to the project.

Comment (jan-law) [reply .vscode/settings.json]: Whoops, I removed that now

**Comment by jan-law** [comment]
CI build shows an NPE when generating a report in `ReportIT`, possibly #666?

```
Feb 02, 2022 6:59:49 PM io.cryostat.core.log.Logger info
INFO: io.cryostat.net.reports.SubprocessReportGenerator processing report
java.util.concurrent.ExecutionException: java.lang.NullPointerException
at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
at io.cryostat.net.reports.SubprocessReportGenerator.generateReportFromFile(SubprocessReportGenerator.java:336)
at io.cryostat.net.reports.SubprocessReportGenerator.main(SubprocessReportGenerator.java:292)
Caused by: java.lang.NullPointerException
at org.openjdk.jmc.flightrecorder.parser.synthetic.SyntheticAttributeExtension$ModuleExportSink.addEvent(SyntheticAttributeExtension.java:206)
at org.openjdk.jmc.flightrecorder.internal.parser.v1.TypeManager$EventTypeEntry.readEvent(TypeManager.java:318)
at org.openjdk.jmc.flightrecorder.internal.parser.v1.TypeManager.readEvent(TypeManager.java:406)
at org.openjdk.jmc.flightrecorder.internal.parser.v1.ChunkLoaderV1.call(ChunkLoaderV1.java:91)
at org.openjdk.jmc.flightrecorder.internal.parser.v1.ChunkLoaderV1.call(ChunkLoaderV1.java:47)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:829)
Feb 02, 2022 6:59:50 PM io.cryostat.core.log.Logger error
SEVERE: Exception thrown
io.cryostat.net.reports.SubprocessReportGenerator$SubprocessReportGenerationException: [6] An unspecified unexpected exception occurred.
at io.cryostat.net.reports.SubprocessReportGenerator.lambda$exec$0(SubprocessReportGenerator.java:163)
at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
at java.base/java.lang.Thread.run(Thread.java:829)

Feb 02, 2022 6:59:50 PM io.cryostat.core.log.Logger error
SEVERE: Exception thrown
java.util.concurrent.ExecutionException: io.cryostat.net.reports.SubprocessReportGenerator$SubprocessReportGenerationException: [6] An unspecified unexpected exception occurred.
at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2022)
at io.cryostat.net.reports.ArchivedRecordingReportCache.get(ArchivedRecordingReportCache.java:90)
at io.cryostat.net.reports.ReportService.get(ReportService.java:57)
at io.cryostat.net.web.http.api.v1.ReportGetHandler.handleAuthenticated(ReportGetHandler.java:123)
at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:90)
at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:65)
at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: io.cryostat.net.reports.SubprocessReportGenerator$SubprocessReportGenerationException: [6] An unspecified unexpected exception occurred.
at io.cryostat.net.reports.SubprocessReportGenerator.lambda$exec$0(SubprocessReportGenerator.java:163)
at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
... 1 more

Feb 02, 2022 6:59:50 PM io.cryostat.core.log.Logger error
SEVERE: HTTP 500: java.util.concurrent.ExecutionException: io.cryostat.net.reports.SubprocessReportGenerator$SubprocessReportGenerationException: [6] An unspecified unexpected exception occurred.
io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
```

**Comment by andrewazores** [comment]
Looks like it. That issue number really turned out almost prophetic...

**Review by andrewazores — APPROVED** [review]

**Pull request merged**